File: src/034/p034.py
from itertools import combinations_with_replacement
import math


DIGIT_FACTORIALS = [math.factorial(d) for d in range(10)]


def solve() -> int:
    total = 0
    for length in range(2, 8):
        for combo in combinations_with_replacement(range(10), length):
            factorial_sum = sum(DIGIT_FACTORIALS[d] for d in combo)
            factorial_sum_digits = sorted([int(d) for d in str(factorial_sum)])
            combo_digits = sorted(combo)
            if factorial_sum_digits == combo_digits:
                total += factorial_sum
    return total

File: src/P034/p034.go
package problem034

import (
	"sort"
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/intutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/sliceutils"
)

func Solve() string {
	factorials := make([]int, 10)
	for d := 0; d < 10; d++ {
		factorials[d] = intutils.Factorial(d)
	}

	answer := 0
	for length := 2; length < 8; length++ {
		for digits := range sliceutils.CombinationsWithReplacement([]int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, length) {
			factorialSum := 0
			for _, d := range digits {
				factorialSum += factorials[d]
			}
			factorialSumDigits := intutils.Digits(strconv.Itoa(factorialSum))
			sort.Ints(factorialSumDigits)

			if sliceutils.Equal(digits, factorialSumDigits) {
				answer += factorialSum
			}
		}
	}
	return strconv.Itoa(answer)
}
